Mathlib/Tactic/Setm.lean

```diff
@@ -0,0 +1,112 @@
+`setm ?a ^ 2 + ?b * ?a + 3 = 28` would introduce `a : ℕ := x + 5` and `b : ℕ := 2 * y + 5` into the
+context and change the goal to `a ^ 2 + b * a + 3 = 28`
+-/
+
+namespace Mathlib.Tactic
+open Lean Parser Tactic Elab Tactic Meta
+
+initialize registerTraceClass `Tactic.setm
+
+/-- This is the core to the `setm` tactic. It takes an expression `e` and pattern `stx` containing
+named holes whose form should match `e`. For each named hole in `stx`, we create a `let` declaration
+of the same name whose value is filled by the corresponding subexpression of `e`. We then return
+the expression `e` with its subexpressions replaced by these named variables. -/
+def setMCore (e : Expr) (stx : TSyntax `term) : TacticM Expr := withMainContext do
+  let (origPattern, mvarIds) ← elabTermWithHoles stx none (←getMainTag) (allowNaturalHoles := true)
+  /- Named holes are by default syntheticOpaque and not assignable, so we change that -/
+  mvarIds.forM fun mvarId => mvarId.setKind .natural
+  trace[Tactic.setm] "origPattern : {← ppExpr origPattern}"
+
+  /- Create new placeholder mvars -/
+  let mvarIdsPairs ← mvarIds.mapM
+    (fun x => return (x, ←mkFreshExprMVar none (userName := (← x.getDecl).userName)))
+  let mvarIdsMap : HashMap MVarId Expr := .ofList mvarIdsPairs
+
+  /- newPattern is the placeholder pattern with a bunch of placeholder mvars -/
+  let newPattern := origPattern.replace fun x => match x with
+    | .mvar x => mvarIdsMap.find? x
+    | _ => none
+
+  /- We let isDefEq match the given pattern -/
+  let mctx ← getMCtx
+  let mdecls := mctx.decls
+  if ←isDefEq e newPattern then
+    /- We iterate over the (old, new) vars -/
+    mvarIdsPairs.forM fun (mvarIdOld, mvarExprNew) => do
+      let mvarIdNew := mvarExprNew.mvarId!
+      match (mdecls.find! mvarIdNew).userName with
+      | Name.anonymous => return ()
+      | name =>
+        let a := mkIdent name
+        let ha := mkIdent $ name.appendBefore "h"
+        let goal ← getMainGoal
+        match (←getExprMVarAssignment? mvarIdNew) with
+        | .some mvarAss =>
+          /-
+          Here mvarIdNew is assigned to mvarAss, and mvarIdOld <-> mvarIdNew, so I
+            (1) let $a := mvarAss -- Here fvarId stores $a (?)
+            (2) have $ha : $a = mvarAss := rfl
+            (3) assign $a to mvarIdOld
+          -/
+          let mvarType ← inferType mvarAss
+          let (fvarId, goal) ← (←goal.define a.getId mvarType mvarAss).intro1P
+          replaceMainGoal [goal]
+          evalTactic (←`(tactic| have $ha : $a = $(←Term.exprToSyntax mvarAss) := rfl))
+          mvarIdOld.assign (.fvar fvarId)
+        | none => throwError "File a bug report!"
+      mvarIdOld.setUserName <| (← mkFreshUserName (← mvarIdOld.getDecl).userName)
+      mvarIdNew.setUserName <| (← mkFreshUserName (← mvarIdNew.getDecl).userName)
+  else
+    throwError f!"setm: pattern is not definitionally equal to the goal."
```

Comment:
Note: `throwError` uses interpolated strings, so we can write things like
```
throwError "setm: {stx} is not definitionally equal to the goal."
```
(You can interpolate any expression that has a `ToMessageData` instance, which is almost everything you'd want to interpolate.)